// crypto-balances/lib/services/counterparty.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, Promise, _, counterparty, req,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Promise = require("bluebird");

  req = Promise.promisify(require("request"));

  _ = require("lodash");

  InvalidResponseError = require("../errors").InvalidResponseError;

  counterparty = function(addr) {
    var url;
    url = "http://xcp.blockscan.com/api2?module=address&action=balance&btc_address=" + addr;
    return req(url, {
      json: true
    }).timeout(4000).cancellable().spread(function(resp, json) {
      var i, ref, results;
      if ((ref = resp.statusCode, indexOf.call((function() {
        results = [];
        for (i = 200; i <= 299; i++){ results.push(i); }
        return results;
      }).apply(this), ref) >= 0) && _.isArray(json.data)) {
        return json.data;
      } else {
        if (_.isObject(json) && json.message === "error") {
          return [];
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      }
    }).map(function(data) {
      return {
        status: "success",
        service: "http://xcp.blockscan.com",
        address: addr,
        quantity: data.balance,
        asset: data.asset
      };
    })["catch"](Promise.TimeoutError, function(e) {
      return [
        {
          status: 'error',
          service: url,
          message: e.message,
          raw: e
        }
      ];
    })["catch"](InvalidResponseError, function(e) {
      return [
        {
          status: "error",
          service: e.service,
          message: e.message,
          raw: e.response
        }
      ];
    });
  };

  module.exports = counterparty;

}).call(this);
