// crypto-balances/lib/services/cryptoid.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, Promise, _, cryptoid, json, req,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Promise = require("bluebird");

  req = Promise.promisify(require("request"));

  _ = require("lodash");

  InvalidResponseError = require("../errors").InvalidResponseError;

  json = [];

  cryptoid = function(addr) {
    var currency, url;
    currency = (function() {
      switch (false) {
        case addr[0] !== 'X':
          return 'dash';
        case addr[0] !== 'C':
          return 'cpc';
        case addr[0] !== 'P':
          return 'ppc';
        case addr[0] !== 'R':
          return 'rby';
        case addr[0] !== 'G':
          return 'grt';
        default:
          return 'blk';
      }
    })();
    url = "http://chainz.cryptoid.info/" + currency + "/api.dws?q=getbalance&a=" + addr;
    return req(url, {
      json: true
    }).timeout(2000).cancellable().spread(function(resp) {
      var i, ref, results;
      if (ref = resp.statusCode, indexOf.call((function() {
        results = [];
        for (i = 200; i <= 299; i++){ results.push(i); }
        return results;
      }).apply(this), ref) >= 0) {
        return {
          status: "success",
          service: "http://chainz.cryptoid.info",
          address: addr,
          quantity: resp.body.toString(),
          asset: currency.toUpperCase()
        };
      } else {
        if (_.isObject(json) && json.message === "error") {
          return [];
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      }
    })["catch"](Promise.TimeoutError, function(e) {
      return [
        {
          status: 'error',
          service: url,
          message: e.message,
          raw: e
        }
      ];
    })["catch"](InvalidResponseError, function(e) {
      return [
        {
          status: "error",
          service: e.service,
          message: e.message,
          raw: e.response
        }
      ];
    });
  };

  module.exports = cryptoid;

}).call(this);
