// crypto-balances/lib/services/doge.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, Promise, _, doge, req,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Promise = require("bluebird");

  req = Promise.promisify(require("request"));

  _ = require("lodash");

  InvalidResponseError = require("../errors").InvalidResponseError;

  doge = function(addr) {
    var url;
    url = "http://dogechain.info/api/v1/address/balance/" + addr;
    return req(url, {
      json: true
    }).timeout(3000).cancellable().spread(function(resp, json) {
      var i, ref, results;
      if (ref = resp.statusCode, indexOf.call((function() {
        results = [];
        for (i = 200; i <= 299; i++){ results.push(i); }
        return results;
      }).apply(this), ref) >= 0) {
        return {
          status: "success",
          service: "https://dogechain.info",
          address: addr,
          quantity: json.balance,
          asset: "DOGE"
        };
      } else {
        if (_.isObject(json) && json.message === "error") {
          return [];
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      }
    })["catch"](Promise.TimeoutError, function(e) {
      return [
        {
          status: 'error',
          service: url,
          message: e.message,
          raw: e
        }
      ];
    })["catch"](InvalidResponseError, function(e) {
      return [
        {
          status: "error",
          service: e.service,
          message: e.message,
          raw: e.response
        }
      ];
    });
  };

  module.exports = doge;

}).call(this);
