// crypto-balances/lib/services/nem.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, Promise, _, converter, nem, req,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Promise = require("bluebird");

  req = Promise.promisify(require("request"));

  _ = require("lodash");

  InvalidResponseError = require("../errors").InvalidResponseError;

  converter = require("./../converter");

  nem = function(addr) {
    var url;
    url = "http://bigalice3.nem.ninja:7890/account/get?address=" + (addr.replace(/-/g, ''));
    return req(url, {
      json: true
    }).timeout(2000).cancellable().spread(function(resp, json) {
      var i, ref, results;
      if (ref = resp.statusCode, indexOf.call((function() {
        results = [];
        for (i = 200; i <= 299; i++){ results.push(i); }
        return results;
      }).apply(this), ref) >= 0) {
        return {
          status: "success",
          service: "http://node.cyber.fund:7890",
          address: addr,
          quantity: converter.toCoin(json.account.balance, "XEM"),
          asset: "XEM"
        };
      } else {
        if (_.isObject(json) && json.message === "error") {
          return [];
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      }
    })["catch"](Promise.TimeoutError, function(e) {
      return [
        {
          status: 'error',
          service: url,
          message: e.message,
          raw: e
        }
      ];
    })["catch"](InvalidResponseError, function(e) {
      return [
        {
          status: "error",
          service: e.service,
          message: e.message,
          raw: e.response
        }
      ];
    });
  };

  module.exports = nem;

}).call(this);
