// crypto-balances/lib/services/openassets.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, Promise, _, oa, req,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Promise = require("bluebird");

  req = Promise.promisify(require("request"));

  _ = require("lodash");

  InvalidResponseError = require("../errors").InvalidResponseError;

  oa = function(addr) {
    var url;
    url = "https://api.coinprism.com/v1/addresses/" + addr;
    return req(url, {
      json: true
    }).timeout(2000).cancellable().spread(function(resp, json) {
      var i, ref, results;
      if ((ref = resp.statusCode, indexOf.call((function() {
        results = [];
        for (i = 200; i <= 299; i++){ results.push(i); }
        return results;
      }).apply(this), ref) >= 0) && json.address === addr && _.isArray(json.assets)) {
        return json.assets;
      } else {
        if (_.isObject(json) && json.Message === "Error" && json.ErrorCode === "InvalidAddress") {
          return [];
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      }
    }).map(function(asset) {
      var assetUrl;
      assetUrl = "https://api.coinprism.com/v1/assets/" + asset.id;
      return req(assetUrl, {
        json: true
      }).timeout(2000).cancellable().spread(function(resp, json) {
        var i, ref, results;
        if (ref = resp.statusCode, indexOf.call((function() {
          results = [];
          for (i = 200; i <= 299; i++){ results.push(i); }
          return results;
        }).apply(this), ref) >= 0) {
          if (_.isNull(json)) {
            return _.merge(asset, {
              symbol: "" + asset.id,
              divisibility: 0
            });
          } else if (json.asset_id === asset.id) {
            return _.merge(asset, {
              symbol: "" + (json.name_short.toUpperCase()),
              divisibility: json.divisibility
            });
          }
        } else {
          throw new InvalidResponseError({
            service: url,
            response: resp
          });
        }
      });
    }).map(function(asset) {
      var balance, quantity;
      balance = parseInt(asset.balance, 10);
      quantity = balance / (Math.pow(10, asset.divisibility));
      return {
        status: "success",
        service: "https://api.coinprism.com",
        address: addr,
        quantity: quantity.toFixed(6),
        asset: asset.symbol
      };
    })["catch"](Promise.TimeoutError, function(e) {
      return [
        {
          status: 'error',
          service: url,
          message: e.message,
          raw: e
        }
      ];
    })["catch"](InvalidResponseError, function(e) {
      return [
        {
          status: "error",
          service: e.service,
          message: e.message,
          raw: e.response
        }
      ];
    });
  };

  module.exports = oa;

}).call(this);
