// crypto-balances/lib/errors.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InvalidResponseError, ReqError,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  ReqError = (function(superClass) {
    extend(ReqError, superClass);

    function ReqError(resp) {
      this.name = "ReqError";
      if (resp.body.data.address.search(/valid address.*required/i >= 0)) {
        this.message = "Invalid address";
      } else {
        this.message = "API error occured";
      }
      this.resp = resp;
    }

    return ReqError;

  })(Error);

  InvalidResponseError = (function(superClass) {
    extend(InvalidResponseError, superClass);

    function InvalidResponseError(opts) {
      this.name = "InvalidResponseError";
      this.message = opts.message || "Server response is invalid";
      this.response = opts.response;
      this.service = opts.service;
    }

    return InvalidResponseError;

  })(Error);

  module.exports = {
    "ReqError": ReqError,
    "InvalidResponseError": InvalidResponseError
  };

}).call(this);
